cli/internal/dataplane/bufferblob.go

```diff
@@ -59,15 +64,182 @@ type BufferEndMetadata struct {
 }
 
 type Container struct {
-	*url.URL
+	lock      *sync.Mutex
+	accessUrl *url.URL
+	getNewUrl func(context.Context) (*url.URL, error)
 }
 
-func clearBit(value int64, pos int) int64 {
-	mask := int64(^(1 << pos))
-	return value & mask
+func NewContainer(accessUrl *url.URL) *Container {
+	return &Container{accessUrl: accessUrl, lock: &sync.Mutex{}}
 }
 
-func (c *Container) GetBlobUrl(blobNumber int64) string {
+func NewContainerFromAccessString(ctx context.Context, accessString string) (*Container, error) {
+	accessUrl, err := ParseBufferAccessUrl(accessString)
+	if err != nil {
+		// Invalid URL, assume it's a file containing the URL
+		return NewContainerFromAccessFile(ctx, accessString)
+	}
+
+	// Valid URL, need to parse the SAS parameters in order to request a new URL later
+	bufferId := path.Base(accessUrl.Path)
+	qv := accessUrl.Query()
+	permissions := qv.Get("sp")
+	writeable := strings.Contains(permissions, "c")
+	st, err := parseSasQueryTimestamp(accessUrl, "st")
+	if err != nil {
+		return nil, err
+	}
+	se, err := parseSasQueryTimestamp(accessUrl, "se")
+	if err != nil {
+		return nil, err
+	}
+	accessTtl := ""
+	d := se.Sub(st)
+	if d > 0 {
+		accessTtl = common.DurationToTimeToLive(d).String()
+	}
+
+	getNewUrl := func(ctx context.Context) (*url.URL, error) {
+		url, err := RequestNewBufferAccessUrl(ctx, bufferId, writeable, accessTtl)
+		if err != nil {
+			return nil, err
+		}
+		return url, nil
+	}
+
+	return newContainer(ctx, accessUrl, getNewUrl)
+}
+
+func NewContainerFromAccessFile(ctx context.Context, filename string) (*Container, error) {
+	getNewUrl := func(ctx context.Context) (*url.URL, error) {
+		url, err := GetBufferAccessUrlFromFile(filename)
+		if err != nil {
+			return nil, err
+		}
+		return url, nil
+	}
+
+	return newContainer(ctx, nil, getNewUrl)
+}
+
+func NewContainerFromBufferId(ctx context.Context, bufferId string, writeable bool, accessTtl string) (*Container, error) {
+	getNewUrl := func(ctx context.Context) (*url.URL, error) {
+		url, err := RequestNewBufferAccessUrl(ctx, bufferId, writeable, accessTtl)
+		if err != nil {
+			return nil, err
+		}
+		return url, nil
+	}
+
+	return newContainer(ctx, nil, getNewUrl)
+}
+
+func newContainer(ctx context.Context, accessUrl *url.URL, getUrl func(context.Context) (*url.URL, error)) (*Container, error) {
+	if accessUrl == nil {
+		url, err := getUrl(ctx)
+		if err != nil {
+			return nil, err
+		}
+		accessUrl = url
+	}
+	c := &Container{
+		lock:      &sync.Mutex{},
+		accessUrl: accessUrl,
+		getNewUrl: getUrl,
+	}
+	return c, nil
+}
+
+// CurrentAccessUrl returns the current access URL without attempting to refresh.
+func (c *Container) CurrentAccessUrl() *url.URL {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.accessUrl
+}
+
+// GetValidAccessUrl returns a valid access URL, refreshing it if necessary.
+func (c *Container) GetValidAccessUrl(ctx context.Context) (*url.URL, error) {
```

Comment:
Did you consider caching the the refresh time on the Container so that this method can be lighter weight in most cases? We could save the parse and lock for most calls

Reply:
Yes, spot on. I had that before doing a heavy refactor. Easy to add back in 👍